### PR TITLE
[TECH] Test page

### PR DESCRIPTION
**Pull request by petemarianne**
here's only test navigation bar, lower buttons + essay page (only desktop version yet)

### CODE REVIEW

File: src/pages/Test/Essay/Essay.jsx [hunk 1]
```diff
@@ -0,0 +1,32 @@
+import React, {Component} from "react";
+import './Essay.scss';
+
+export default class Essay extends Component {
```
Comment (tensenrey) [inline src/pages/Test/Essay/Essay.jsx]: why are you using the class component ?

File: src/pages/Test/Grammar/Grammar.jsx [hunk 2]
```diff
@@ -0,0 +1,289 @@
+import React from "react";
+import './Grammar.scss';
+
+const Grammar = () => {
+  const grammarTasks = [
```
Comment (tensenrey) [inline src/pages/Test/Grammar/Grammar.jsx]: create new file structure /src/testData/grammarTasks.js

File: src/pages/Test/Grammar/Grammar.jsx [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+          isRight: false
+        },
+        {
+          option: 'have to',
+          isRight: false
+        }
+      ]
+    }
+  ];
+  let number = 0;
```
Comment (tensenrey) [inline src/pages/Test/Grammar/Grammar.jsx]: please give the variable a logical name

File: src/pages/Test/Listening/Listening.jsx [hunk 4]
```diff
@@ -0,0 +1,248 @@
+import React from "react";
+import "./Listening.scss";
+const Listening = () => {
+
+  const grammarTasks = [
```
Comment (tensenrey) [inline src/pages/Test/Listening/Listening.jsx]: create new file structure /src/testData/grammarTasks.js

File: src/pages/Test/Speaking/Speaking.jsx [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+        }, 500);
+      });
+
+
+      mediaRecorder.addEventListener("dataavailable", function (event) {
+        const key = Math.random();
+        voice.push({key: key, time: parseInt((Date.now() - startTime) / 1000), data: event.data});
+        url = URL.createObjectURL(event.data);
+
+        let obj = document.createElement('audio');
```
Comment (tensenrey) [inline src/pages/Test/Speaking/Speaking.jsx]: please give the variable a logical name

File: src/pages/Test/Speaking/Speaking.jsx [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+        document.getElementById('audio').appendChild(obj);
+
+      });
+
+      document.getElementById('stop').addEventListener('click', function () {
+        mediaRecorder.stop();
+        clearTimeout(startTimeId);
+      });
+
+
```
Comment (tensenrey) [inline src/pages/Test/Speaking/Speaking.jsx:8]: remove empty lines

File: src/pages/Test/Speaking/Speaking.jsx [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+    });
+
```
Comment (tensenrey) [inline src/pages/Test/Speaking/Speaking.jsx]: remove empty line

File: src/pages/Test/Test.jsx [hunk 8]
```diff
@@ -0,0 +1,156 @@
+import React, {Component} from "react";
+import './Test.scss';
+import Layout from "../../components/layout/Layout";
+import Speaking from "./Speaking/Speaking";
+import Essay from "./Essay/Essay";
+import Grammar from "./Grammar/Grammar";
+import Listening from "./Listening/Listening";
+import Button from '@material-ui/core/Button';
+
+export default class Test extends Component {
```
Comment (tensenrey) [inline src/pages/Test/Test.jsx]: why are you using the class component ?

File: src/pages/Test/Test.scss [hunk 9]
```diff
@@ -0,0 +1,90 @@
+.wrapper1 {
```
Comment (tensenrey) [inline src/pages/Test/Test.scss]: rename this please

File: src/pages/Test/Test.scss [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+  height: 90px; /* Задаем высоту обертки */
+
+  .wrapper1 :last-child {
+    margin-right: 0; /* Убираем поле последнего div-а */
+  }
+}
+
+.bl {
+  float: left; /* Задаем обтекание */
+  line-height: 100%; /* Высота строки +  верт. центрирования текста */
```
Comment (tensenrey) [inline src/pages/Test/Test.scss]: write comments in English

File: src/App.js [hunk 11]
```diff
@@ -23,8 +24,8 @@ const App = () => {
         <PrivateRoute path="/" exact><Main /></PrivateRoute>
         <PrivateRoute path="/jwt" ><DemoJWT /></PrivateRoute>
         <PrivateRoute path="/edittests" ><EditTests /></PrivateRoute>
+        <Route path='/test'><Test /></Route>
```
Comment (mtalstykh-exadel) [inline src/App.js]: why the route is not private?

File: src/assets/images/play.svg [hunk 12]
```diff
@@ -0,0 +1,3 @@
+<svg width="40" height="40" viewBox="0 0 40 40" fill="none" xmlns="http://www.w3.org/2000/svg">
```
Comment (mtalstykh-exadel) [inline src/assets/images/play.svg]: use material icons instead of these SVGs

File: src/pages/Test/Essay/Essay.jsx [hunk 13]
```diff
@@ -0,0 +1,38 @@
+import React, { useState } from "react";
+import "./Essay.scss";
+
+const Essay = () => {
+  const [value, setValue] = useState("");
```
Comment (mtalstykh-exadel) [inline src/pages/Test/Essay/Essay.jsx]: value... of what? variable name is not good

File: src/pages/Test/Essay/Essay.scss [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+.essay-input {
+  width: 66%;
+  min-height: 220px;
+  border: 1px solid #bdbdbd;
+  margin-top: 20px;
+  resize: none;
+  box-sizing: border-box;
+}
+
+.essay-count {
```
Comment (mtalstykh-exadel) [inline src/pages/Test/Essay/Essay.scss]: seems it is not what you wanted to write

image

File: src/pages/Test/Grammar/Grammar.jsx [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+            type="radio"
+            name={"group-" + numberOfTheQuestion}
+            value={item2.option}
+          />
+          <label> {item2.option}</label>
+        </div>
+      );
+    });
+    return (
+      <div key={numberOfTheQuestion}>
```
Comment (mtalstykh-exadel) [inline src/pages/Test/Grammar/Grammar.jsx]: please, format the code. So hard to read =(

File: src/pages/Test/Grammar/Grammar.jsx [hunk 16]
```diff
@@ -0,0 +1,46 @@
+import React from "react";
+import PropTypes from "prop-types";
+import "./Grammar.scss";
+
+const Grammar = ({ tasks }) => {
+  let numberOfTheQuestion = 0;
+  const questions = tasks.map((item1) => {
```
Comment (mtalstykh-exadel) [inline src/pages/Test/Grammar/Grammar.jsx]: why "item1"? why not a "question"?

File: src/pages/Test/Essay/Essay.jsx [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+
+  const handleChange = (event) => {
+    setValue(event.target.value);
+  };
+
+  return (
+    <>
+      <div className="step-description">Write an essay on a given topic</div>
+      <div className="essay-topic">Essay Topic</div>
+      <textarea
```
Comment (mtalstykh-exadel) [inline src/pages/Test/Essay/Essay.jsx]: https://stackoverflow.com/questions/46393703/text-area-in-material-ui

File: src/pages/Test/Speaking/Speaking.jsx [hunk 18]
```diff
@@ -0,0 +1,42 @@
+import React, { useState } from "react";
+import micOn from "../../../assets/images/micOn.svg";
+import micOff from "../../../assets/images/micOff.svg";
```
Comment (mtalstykh-exadel) [inline src/pages/Test/Speaking/Speaking.jsx]: use material icons

File: src/pages/Test/Speaking/Speaking.scss [hunk 19]
```diff
@@ -0,0 +1,43 @@
+.img img {
+  width: 60px;
+  padding-bottom: 40px;
+  cursor: pointer;
+}
+
+#start {
```
Comment (mtalstykh-exadel) [inline src/pages/Test/Speaking/Speaking.scss]: do not use IDs if you can use classes

File: src/pages/Test/Speaking/Speaking.scss [hunk 20]
```diff
@@ -0,0 +1,43 @@
+.img img {
```
Comment (mtalstykh-exadel) [inline src/pages/Test/Speaking/Speaking.scss]: that a bad name

File: src/pages/Test/Grammar/Grammar.scss [hunk 21]
```diff
@@ -0,0 +1,22 @@
...
+    font-weight: 400;
+  }
+}
+
+.option {
+  padding-top: 13px;
+  text-align: left;
+  font-size: 18px;
+  font-weight: 400;
+}
```
Comment (snaf-trip) [inline src/pages/Test/Grammar/Grammar.scss]: that style will affect all .option elements across whole project

File: src/pages/Test/Speaking/Speaking.jsx [hunk 22]
```diff
@@ -0,0 +1,79 @@
...
+      <div className="speaking-topic">Speaking Topic</div>
+      <div className="audio-speaking-timer" id="speaking-timer">
+        5:00
+      </div>
+      <div className={invisible === "off" ? "microphone" : "microphone off"}>
+        {invisible === "off" ? (
+          <MicIcon
+            alt="microOn"
+            className="microphone-item"
+            style={{ color: "white" }}
```
Comment (snaf-trip) [inline src/pages/Test/Speaking/Speaking.jsx:47]: use SCSS to style components

File: src/pages/Test/Speaking/Speaking.jsx [hunk 23]
```diff
@@ -0,0 +1,79 @@
...
+              onRecAudio();
+              speakingTimerHandler(true);
+              SpeakingTimerHandker();
+            }}
+          />
+        ) : (
+          <MicOffIcon
+            alt="microOff"
+            className="microphone-item"
+            style={{ color: "white" }}
```
Comment (snaf-trip) [inline src/pages/Test/Speaking/Speaking.jsx]: use SCSS to style components

File: src/pages/Test/Speaking/Speaking.jsx (hunk 22 above)
Comment (snaf-trip) [inline src/pages/Test/Speaking/Speaking.jsx]: use SCSS to style components

File: src/pages/Test/player/player.jsx [hunk 24]
```diff
@@ -0,0 +1,119 @@
...
+          {audioCurrent === 0 ? "0:00" : audioCurrent}/
+          {checkTime(audioDuration)}
+        </div>
+        <div className="progress-container">
+          <audio id="audio-player" src={src}></audio>
+          <div
+            style={{ width: progressPercent + "%" }}
+            className="progress-line"
+          ></div>
+          <div className="progress"></div>
```
Comment (snaf-trip) [inline src/pages/Test/player/player.jsx]: why do I need `< /div>` if there is nothing in it

File: src/pages/Test/player/player.scss [hunk 25]
```diff
@@ -0,0 +1,124 @@
+.player {
+  position: relative;
+  margin: 0 auto;
+  width: 66%;
+  height: 55px;
+  background: rgba(62, 109, 215, 0.15);
```
Comment (snaf-trip) [inline src/pages/Test/player/player.scss:6]: use background-color

File: src/services/timer.js [hunk 26]
```diff
@@ -0,0 +1,39 @@
+let audioDuration = 0;
+let timeMinut;
+let timer;
+
+const speakingTimerHandler = (timerOn) => {
+  const speakingTimerDomElement = document.getElementById("speaking-timer");
+  timeMinut = 299;
```
Comment (snaf-trip) [inline src/services/timer.js]: is like a  299 minutes. Rename pls

**Review by fshovchko — DISMISSED** [review]

File: src/components/testLevelSelector/testLevelSelectorItem/testLevelSelectorItem.jsx [hunk 27]
```diff
@@ -9,9 +10,11 @@ const TestLevelsSelectorItem = ({ title, description }) => {
     <div className='test-level-selector-item'>
       <div className="title">{title}</div>
       <div className="description">{description}</div>
-      <Button variant="contained" color="primary" className='btn'>
+      <Link to="/test" className='btn'>
```
Comment (fshovchko) [inline src/components/testLevelSelector/testLevelSelectorItem/testLevelSelectorItem.jsx]: You can write link button like this:
https://github.com/mtalstykh-exadel/team4_frontend/blob/a146ae45a8b03735d6d53da1b924d024244cc9fd/src/components/header/header.js#L70-L76

File: src/pages/Test/Speaking/Speaking.scss [hunk 28]
```diff
@@ -0,0 +1,38 @@
+.microphone {
+  display: flex;
+  align-items: center;
+  justify-content: center;
+  width: 60px;
```
Comment (fshovchko) [inline src/pages/Test/Speaking/Speaking.scss:5]: you can move 60px for width and height to variable

File: src/pages/Test/player/player.scss [hunk 29]
```diff
@@ -0,0 +1,124 @@
...
+  display: flex;
+  align-items: flex-start;
+}
+
+.player-button {
+  margin-top: 8px;
+  margin-left: 10px;
+  margin-right: 5px;
+  width: 40px;
+  height: 40px;
```
Comment (fshovchko) [inline src/pages/Test/player/player.scss:17]: You can use variables for repeating values

**Review by fshovchko — DISMISSED** [review]

File: src/services/test-timer.js [hunk 30]
```diff
@@ -0,0 +1,21 @@
...
+          seconds = "0" + seconds;
+        }
+        timeSeconds--;
+        testTimerDomElement.textContent = `${Math.floor(minutes)}:${seconds}`;
+      }
+    }, 1000);
+  }
+};
+
+export { testTimerHandler };
```
Comment (fshovchko) [inline src/services/test-timer.js:21]: why do you need second timer?

File: src/services/speaking-timer.js [hunk 31]
```diff
@@ -0,0 +1,39 @@
...
+    } else {
+      if (seconds < 10) {
+        seconds = "0" + seconds;
+      }
+      ++audioDuration;
+      --timeSeconds;
+      element.textContent = `${Math.floor(minutes)}:${seconds}`;
+    }
+  }, 1000);
+};
```
Comment (fshovchko) [inline src/services/speaking-timer.js:37]: Check out the solution for timer I have. It takes seconds and returns time in format h : m :
image

File: package.json [hunk 32]
```diff
@@ -9,6 +9,7 @@
     "@testing-library/jest-dom": "^5.14.1",
     "@testing-library/react": "^11.2.7",
     "@testing-library/user-event": "^12.8.3",
+    "@types/dom-mediacapture-record": "^1.0.10",
```
Comment (mtalstykh-exadel) [inline package.json]: why do you need it?

File: src/components/testLevelSelector/testLevelSelectorItem/testLevelSelectorItem.scss [hunk 33]
```diff
@@ -18,5 +18,6 @@
     width: 180px;
     height: 40px;
     margin: 20px 0 40px;
+    text-decoration: none;
```
Comment (mtalstykh-exadel) [inline src/components/testLevelSelector/testLevelSelectorItem/testLevelSelectorItem.scss:21]: change naming of the file

File: src/services/speaking-timer.js (hunk 31 above)
Comment (tensenrey) [reply src/services/speaking-timer.js:37]: it is not possible to read

File: src/pages/Test/Speaking/Speaking.jsx [hunk 34]
```diff
@@ -0,0 +1,67 @@
...
+        {invisible === "off" ? (
+          <MicIcon
+            alt="microOn"
+            className="microphone-item"
+            name="onButton"
+            onClick={() => {
+              setInvisible("on");
+              onRecAudio();
+              speakingTimerHandler(true);
+              SpeakingTimerHandler();
```
Comment (mtalstykh-exadel) [inline src/pages/Test/Speaking/Speaking.jsx]: two functions with almonst an identical naming

File: src/pages/Test/Speaking/Speaking.scss [hunk 35]
```diff
@@ -0,0 +1,38 @@
...
+  height: 60px;
+  margin: 0 auto;
+  margin-bottom: 60px;
+  background-color: #3e6dd7;
+  border-radius: 50%;
+  cursor: pointer;
+}
+
+.off {
```
Comment (mtalstykh-exadel) [inline src/pages/Test/Speaking/Speaking.scss]: what "off"? strange naming

File: src/pages/Test/Test.scss [hunk 36]
```diff
@@ -0,0 +1,89 @@
...
+  .buttons-wrapper :last-child {
+    margin-right: 0;
+  }
+}
+
+.test-buttons {
+  display: flex;
+  justify-content: space-between;
+}
+
```
Comment (mtalstykh-exadel) [inline src/pages/Test/Test.scss:68]: use style nesting

Comment (tensenrey) [reply src/pages/Test/Test.scss:68]: it is already used here

File: src/components/testLevelSelector/testLevelSelector.jsx [hunk 37]
```diff
@@ -1,6 +1,6 @@
 /* eslint-disable max-len */
 import React from 'react';
-import TestLevelsSelectorItem from './testLevelSelectorItem/testLevelSelectorItem';
+import TestLevelsSelectorItem from './testLevelSelectorItem/selector-item/item.jsx';
```
Comment (mtalstykh-exadel) [inline src/components/testLevelSelector/testLevelSelector.jsx]: why???

File: src/App.js [hunk 38]
```diff
@@ -9,8 +9,10 @@ import Main from './pages/main/main';
 import DemoJWT from './pages/jwt/jwt.jsx';
 import Page404 from "./pages/Page404/Page404";
 import EditTests from './pages/editTests/editTests';
+import Test from "./pages/Test/Test";
 import '../src/styles/modal.scss';
 
+
```
Comment (mtalstykh-exadel) [inline src/App.js]: remove it

**Review by snaf-trip — DISMISSED** [review]

File: src/components/testLevelSelector/testLevelSelector.jsx [hunk 39]
```diff
@@ -51,4 +51,4 @@ const TestLevelsSelector = () => {
   );
 };
 
-export default TestLevelsSelector;
+export default LevelSelector;
```
Comment (mtalstykh-exadel) [inline src/components/testLevelSelector/testLevelSelector.jsx]: TestLevelsSelector

File: src/components/testLevelSelector/testLevelSelector.jsx [hunk 40]
```diff
@@ -1,6 +1,6 @@
 /* eslint-disable max-len */
 import React from 'react';
-import TestLevelsSelectorItem from './testLevelSelectorItem/testLevelSelectorItem';
+import TestLevelsSelectorItem from './testLevelSelectorItem/testLevelSelectorItem.jsx';
```
Comment (mtalstykh-exadel) [inline src/components/testLevelSelector/testLevelSelector.jsx]: no need to add ".jsx"

File: src/pages/Test/Speaking/Speaking.jsx [hunk 41]
```diff
@@ -0,0 +1,67 @@
+import React, { useState } from "react";
+import { offRecAudio, onRecAudio } from "../../../services/voice-recorder.js";
+import { speakingTimerHandler } from "../../../services/speaking-timer.js";
+import MicOffIcon from "@material-ui/icons/MicOff";
+import MicIcon from "@material-ui/icons/Mic";
+import Player from "../player/player.jsx";
```
Comment (fshovchko) [inline src/pages/Test/Speaking/Speaking.jsx]: remove file extensions

**Review by fshovchko — DISMISSED** [review]

**Review by fshovchko — APPROVED** [review]

**Review by mtalstykh-exadel — APPROVED** [review]

**Pull request merged**